### PR TITLE
Emulator Support and Improvements - In Progress

### PR DESCRIPTION
**Pull request by rupeshkumar22**
Till now this PR successfully downloads and installs the emulator with the help of SDK manager. We can launch the emulator after SDK Download at startup with `Run In Emulator` option. No manual unpack/download/extraction is being used. The emulator is getting downloaded after acceptance of licence because of an issue-
https://github.com/processing/processing-android/issues/696#issuecomment-1193065934

Also, download emulator process's stream handling was not working properly in SwingWorker Task of SDKDownloaderjava. So I shifted to the end of the download method of SDKAndroid.java

Some To Do Tasks-

- Use Streampump to redirect download stream on PDE, currently, a stream handler is being used to redirect stream on the IDE's console.

- Show execution task in the progress dialog.
- Code format, proper indentation, and removal of debugging logs(print statements)

An outcome of Google Summer of Code 2022
Author - Rupesh Kumar @rupeshkumar22
Mentor - Aditya Rana @ranaaditya
Project Lead(Processing For Android) - Andrés Colubri @codeanticode

### CODE REVIEW

File: mode/src/processing/mode/android/RedirectStreamHandler.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public void run() {
+      try {
+        BufferedReader reader = new BufferedReader(new InputStreamReader(input));
+        String line;
+        while ((line = reader.readLine()) != null) {
+            output.println(line);
+        }
+      } catch (IOException ioe) {
```
Comment (ranaaditya) [inline mode/src/processing/mode/android/RedirectStreamHandler.java:33]: rename ioe to exception or something with better meaning !

please dont use abbreviations for variable names :)

File: mode/src/processing/mode/android/AndroidSDK.java [hunk 2]
```diff
@@ -501,15 +502,26 @@ static public AndroidSDK download(final Frame editor, final AndroidMode androidM
...
-      AndroidUtil.showMessage(AndroidMode.getTextString("android_sdk.dialog.sdk_installed_title"), msg);      
+     sdk.acceptLicenses();
+     File sdkManager = findCliTool(new File(sdk.getCommandLineToolsFolder(), "bin"), "sdkmanager");
+     try {
+    	 String[] cmdToDownloadEmu = {sdkManager.getCanonicalPath(),"emulator"};
+      Process process = Runtime.getRuntime().exec(cmdToDownloadEmu, null, sdk.getFolder());
+      new RedirectStreamHandler(new PrintWriter(System.out, true), process.getInputStream());
+      new RedirectStreamHandler(new PrintWriter(System.out, true), process.getErrorStream());
+      int emulatorDownloadResultCode = process.waitFor();
+      System.out.println("*-*-*-*itsmylog:EmulatorResultCode:"+emulatorDownloadResultCode);
```
Comment (ranaaditya) [inline mode/src/processing/mode/android/AndroidSDK.java:513]: please change the log messages and dont use special signs in logs !

**Review by ranaaditya — CHANGES_REQUESTED** [review]

File: mode/src/processing/mode/android/RedirectStreamHandler.java (hunk 1 above)
Comment (rupeshkumar22) [reply mode/src/processing/mode/android/RedirectStreamHandler.java:33]: Yes, it's important to use proper variable names, I will do that. Also, I'm planning to remove this temporary stream handler file for testing and use PDE's console to print the stream.

File: mode/src/processing/mode/android/AndroidSDK.java (hunk 2 above)
Comment (rupeshkumar22) [reply mode/src/processing/mode/android/AndroidSDK.java:513]: Sure, I will remove these logs for testing in the next commits, Thanks for pointing Aditya

**Comment by rupeshkumar22** [comment]
Some more information-
The console logs, I am sharing in this file
[EmuDonwloadlogsPR703.txt](https://github.com/processing/processing-android/files/9175270/EmuDonwloadlogsPR703.txt)
I observe that when `sdkmanager emulator` command is being executed, It downloads a patch `SDK Patch Applier v4` before the emulator download and install(`emulator-windows_x64`). That can be a reason that emulator was not identified earlier with the manual download and unpacking of the emulator. But anyway now we are using `sdkmanager` to download and install the emulator which works well and the emulator package is identifiable by the `sdkmanager` and `avdmanager` to execute further processes.

**Comment by codeanticode** [comment]
@rupeshkumar22 @ranaaditya great progress, thanks for keep updating the PR. Is it ready for merging or you'd rater push a few more commits?

**Comment by rupeshkumar22** [comment]
@codeanticode @ranaaditya Currently, I'm testing the emulator after the download. Since the emulator downloading is a part of SDK download process, Testing after downloading is little time consuming. I will add final commits today after testing. Thanks!

**Comment by ranaaditya** [comment]
> @rupeshkumar22 @ranaaditya great progress, thanks for keep updating the PR. Is it ready for merging or you'd rater push a few more commits?

@codeanticode Rupesh need to add some more commits to make it working, so we can merge this PR after that and testing as well.

Thanks 1

**Comment by ranaaditya** [comment]
> @codeanticode @ranaaditya Currently, I'm testing the emulator after the download. Since the emulator downloading is a part of SDK download process, Testing after downloading is little time consuming. I will add final commits today after testing. Thanks!

yeah, totally fine, take your time Rupesh but make sure no other component break due to this update i.e. test your code before pushing directly to PR.

thanks !

**Comment by rupeshkumar22** [comment]
A stable package is there with the name - `SDK Patch Applier v4` in `repository2-1.xml` It seems that it needs to be installed to make the emulator package work. In that case, we don't need to use the process builder to install an emulator. Also, there would not be a major change and the PR would be more stable with the current download and unpack methods. 
SDK manager can be used to install the emulator which is shown by this PR but there may be some workaround to show progress in the download dialog from the executed task's stream. For now, I may opt to use the current download and unpack methods if they fit well and if the emulator works with downloading the v4 patch package in the same way. Testing is in progress.

**Comment by rupeshkumar22** [comment]
The emulator was not getting identified by the `sdkmanager` because the package.xml file is not there in the extracted folder in the SDK. Also, package.xml file is not part of the downloaded zip files in the temp folder but these are included in extracted SDK packages. If I include package.xml file in emulator package then it works just fine, but by default package.xml file is not included after extraction which is the reason it's not getting identified by the `sdkmanager` after download and extract. @codeanticode @ranaaditya Any information on how package.xml is getting included in extracted sdk packages such as build tools, platform, and USB driver ([logic here](https://github.com/processing/processing-android/blob/475224e1d0572944f34306b9bb6dd4d1df8a92d0/mode/src/processing/mode/android/SDKDownloader.java#L179)) but are not there in donwloaded zip files? I'm looking forward to it.
Thanks

**Comment by ranaaditya** [comment]
> The emulator was not getting identified by the `sdkmanager` because the package.xml file is not there in the extracted folder in the SDK. Also, package.xml file is not a part of the downloaded zip files in the temp folder but these are included in extracted SDK packages. If I include package.xml file in emulator package then it works just fine, but by default package.xml file is not included after extraction which is the reason it's not getting identified by the `sdkmanager` after download and extract. @codeanticode @ranaaditya Any information on how package.xml is getting included in extracted sdk packages such as build tools, platform, and USB driver ([logic here](https://github.com/processing/processing-android/blob/475224e1d0572944f34306b9bb6dd4d1df8a92d0/mode/src/processing/mode/android/SDKDownloader.java#L179)) but are not there in donwloaded zip files? I'm looking forward to exploring more. Thanks

@rupeshkumar22 I am not exactly sure about it, need to look at it, but I think @codeanticode wrote it's implementation and can tell in brief. If Andres is busy then I will look into it tommorow and we will debug it more precisely.

Thanks !

**Comment by ranaaditya** [comment]
> The emulator was not getting identified by the `sdkmanager` because the package.xml file is not there in the extracted folder in the SDK. Also, package.xml file is not a part of the downloaded zip files in the temp folder but these are included in extracted SDK packages. If I include package.xml file in emulator package then it works just fine, but by default package.xml file is not included after extraction which is the reason it's not getting identified by the `sdkmanager` after download and extract. @codeanticode @ranaaditya Any information on how package.xml is getting included in extracted sdk packages such as build tools, platform, and USB driver ([logic here](https://github.com/processing/processing-android/blob/475224e1d0572944f34306b9bb6dd4d1df8a92d0/mode/src/processing/mode/android/SDKDownloader.java#L179)) but are not there in donwloaded zip files? I'm looking forward to exploring more. Thanks

@rupeshkumar22 so there can be 2 conditions primarily here primarily:
1. package.xml file is there in downloaded zip file but our unpacking function is not including it while unpacking it (which I guess our case is here, not sure)
2. package.xml is not there in downloaded zip file (bigger issue)

So, to check in which case we lie, you need manually unzip the downloaded zip file and verify its contents with the contents of the unpacking function. I dont think there is any other way pssoible to debug it except doing it manually, I did the same thing but for debugging the libraries included in the fat jar file of processing-android long back.

To know in depth how unpacking / unzipping a zip file works you can check out few resources here:
- [unzippping function implementation in java - it has many *.xml files in zip content](https://www.journaldev.com/960/java-unzip-file-example)
- [Extrating xml file from zip file](https://stackoverflow.com/questions/12766171/how-to-read-from-an-xml-within-a-zip-file-in-java)
- [Read zipped xml file](https://gist.github.com/mtakemiya/1329618)
- [Parallely processing large xml files (just for knowledge of handling large xml files) - advance reading, read it if you have time](https://blog.clairvoyantsoft.com/how-to-parallelly-process-large-xml-files-in-java-9ff3a2d32e90)

Debug the issue, if you still getting it or unable to resolve it then we will meet and debug it together.

Thanks,
happy coding !

**Comment by rupeshkumar22** [comment]
@ranaaditya it's the second condition here, after commenting on the code of deletion of temp folder [here](https://github.com/processing/processing-android/blob/55bfe5891be4b57b091553882eaf7a6062cd8876/mode/src/processing/mode/android/SDKDownloader.java#L186), I was able to check the required zip file I confirm that no package.xml files were there in donwloaded zip files after mannual extraction on windows. It seems that package.xml are being created during or after extraction in the program. To resolve emulator unidentified issue by sdkmanager we can add a process to create package.xml files after download and extraction. But it is necessary to check how package.xml has been found in some packages of sdk set by the mode. Thanks for the references I'm looking into these and trying to resolve this main issue.

**Comment by ranaaditya** [comment]
> @ranaaditya it's the second condition here, after commenting on the code of deletion of temp folder [here](https://github.com/processing/processing-android/blob/55bfe5891be4b57b091553882eaf7a6062cd8876/mode/src/processing/mode/android/SDKDownloader.java#L186), I was able to check the required zip file I confirm that no package.xml files were there in donwloaded zip files after mannual extraction on windows. It seems that package.xml are being created during or after extraction in the program. To resolve emulator unidentified issue by sdkmanager we can add a process to create package.xml files after download and extraction. But it is necessary to check how package.xml has been found in some packages of sdk set by the mode. Thanks for the references I'm looking into these and trying to resolve this main issue.

Hi @rupeshkumar22 , got your issue, checkout [this thread](https://stackoverflow.com/questions/62145223/how-to-add-sdk-files-manually-that-needs-a-package-xml-in-android-studio) it shows how package.xml is being generated by android studio after downloading sdk manager.

**Comment by ranaaditya** [comment]
> @ranaaditya it's the second condition here, after commenting on the code of deletion of temp folder [here](https://github.com/processing/processing-android/blob/55bfe5891be4b57b091553882eaf7a6062cd8876/mode/src/processing/mode/android/SDKDownloader.java#L186), I was able to check the required zip file I confirm that no package.xml files were there in donwloaded zip files after mannual extraction on windows. It seems that package.xml are being created during or after extraction in the program. To resolve emulator unidentified issue by sdkmanager we can add a process to create package.xml files after download and extraction. But it is necessary to check how package.xml has been found in some packages of sdk set by the mode. Thanks for the references I'm looking into these and trying to resolve this main issue.

@rupeshkumar22 package.xml file will not be in the downloaded file, android studio generates it from its own after extracting the contents of the zip file, in the above thread I mentioned you can see how studio is also generating package.xml file from the extracted zip file in the temp folder, so yes you are right we need to add a routine which will create package.xml file for us to make emulator being identified by the sdk manager.

**Comment by ranaaditya** [comment]
Screenshot 2022-07-26 at 9 17 00 PM
Screenshot 2022-07-26 at 9 17 18 PM


@rupeshkumar22 for your referrence this is the content of package.xml file for me.

path: "~/Android/sdk/emulator/package.xml"

**Comment by rupeshkumar22** [comment]
Amazing!  Yes, I will create a routine to check the `package.xml` file. The flow will go like this-

1. User Click the `Download New SDK` option.
2. Zip package files get downloaded in tempFolder.
3. Zip package files get copied and extracted to the SDK folder.
4. Check if package.xml files are present in the required SDK folders(i.e build tools, platforms, emulator, etc) - `NEW STEP`
5. If package.xml is not present in the packages, it will create these files. - `NEW STEP`
6. Download SDK completes and User Accepts License.

After this, I am sure there would not be an issue in creating and launching the emulator.
Thanks for sharing these useful links. @ranaaditya

**Comment by ranaaditya** [comment]
> Amazing! Yes, I will create a routine to check the `package.xml` file. The flow will go like this-
> 
> 1. User Click the `Download New SDK` option.
> 2. Zip package files get downloaded in `tempFolder`.
> 3. Zip package files get copied and extracted to the SDK folder. The `tempFolder` gets deleted.
> 4. Check if package.xml files are present in the required SDK folders(i.e build tools, platforms, emulator, etc). - `NEW STEP`
> 5. If package.xml is not present in the packages, it will create these files. - `NEW STEP`
> 6. Download SDK completes and User Accepts License.
> 
> After this, I am sure there would not be an issue in creating and launching the emulator. Thanks for sharing these useful links. @ranaaditya

sounds interesting, do let me know @rupeshkumar22 if you get stuck somewhere in between :+1: 

thanks !

**Comment by codeanticode** [comment]
Good finding about package.xml. Seems like each component in the SDK has it own copy of package.xml, these copies all have the same structure, for example this is the one I have in the emulator, as part of the SDK downloaded by Android Studio (I removed the text of the license agreement for clarity):

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?><ns2:repository xmlns:ns2="http://schemas.android.com/repository/android/common/02" xmlns:ns3="http://schemas.android.com/repository/android/common/01" xmlns:ns4="http://schemas.android.com/repository/android/generic/01" xmlns:ns5="http://schemas.android.com/repository/android/generic/02" xmlns:ns6="http://schemas.android.com/sdk/android/repo/addon2/01" xmlns:ns7="http://schemas.android.com/sdk/android/repo/addon2/02" xmlns:ns8="http://schemas.android.com/sdk/android/repo/addon2/03" xmlns:ns9="http://schemas.android.com/sdk/android/repo/repository2/01" xmlns:ns10="http://schemas.android.com/sdk/android/repo/repository2/02" xmlns:ns11="http://schemas.android.com/sdk/android/repo/repository2/03" xmlns:ns12="http://schemas.android.com/sdk/android/repo/sys-img2/03" xmlns:ns13="http://schemas.android.com/sdk/android/repo/sys-img2/02" xmlns:ns14="http://schemas.android.com/sdk/android/repo/sys-img2/01"><license id="android-sdk-license" type="text">Terms and Conditions

This is the Android Software Development Kit License Agreement

[...]

January 16, 2019</license><localPackage path="emulator" obsolete="false"><type-details xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="ns5:genericDetailsType"/><revision><major>31</major><minor>2</minor><micro>10</micro></revision><display-name>Android Emulator</display-name><uses-license ref="android-sdk-license"/></localPackage></ns2:repository>
```

So should not be hard to generate manually. 

However, I also noticed that in the SDK downloaded by the Android mode, the package.xml file does exist for platform-tools, build-tools, and platform. So why is missing for the emulator?

**Comment by codeanticode** [comment]
@rupeshkumar22 @ranaaditya  Any updates about this? After reading this thread again, my impression is that the emulator can get only get correctly installed with the sdkmanager after the licenses for the other components have been accepted. In this way, the installed emulator contains a valid package.xml file and it's recognized by the sdk. 

Then you started to consider downloading the emulator using the mode's SDK downloader, and creating the package.xml file manually. Are you still trying this approach?

**Comment by rupeshkumar22** [comment]
@codeanticode @ranaaditya Due to the use of emulator installation with sdkmanager outside the SDKDownloader's Swing worker task, It was not getting loaded as [AndroidSDK.load](https://github.com/processing/processing-android/blob/55bfe5891be4b57b091553882eaf7a6062cd8876/mode/src/processing/mode/android/SDKDownloader.java#L198) was not getting used outside SDKDownloader where the emulator getting installed in the commits of this PR. If I use Runtime Execution task inside Download Swing Worker Task, It was not redirecting streams properly, So we can't see the progress. One more thing which would be pending is that we will have to reflect the progress from the Runtime Execute task streams in the download dialog box. We have to filter the progress from the runtime task to reflect on the download dialog. 

Due to all these. For now, I prefer to use the mode's way to download the emulator, and package.xml will be included if not present which resolves the emulator package not found the issue. Also, the current download dialog will show the emulator progress through the total size calculated with [urlholder](https://github.com/processing/processing-android/blob/55bfe5891be4b57b091553882eaf7a6062cd8876/mode/src/processing/mode/android/SDKDownloader.java#L422) in the usual way. For now, I am reverting some previous commits and adding enhancement in the current mode's way of downloading an emulator and adding package.xml manually with a task that would be more stable. In this way, we will be able to at least download and launch the emulator. We can definitely move to `sdkmanager` for installation afterward.

**Comment by rupeshkumar22** [comment]
@codeanticode @ranaaditya I am using something like the following in SDKDownloader to create package.xml if it does not exists inside the downloaded emulator in sdk-
```
File emulatorPackageXMLTemplate = getContentFile("templates/" + EMULATOR_PACKAGE_TEMPLATE);
File packageFileforEmulatorinSDK = new File(emulatorFolder.getAbsolutePath()+"/"+"package.xml");
if(!packageFileforEmulatorinSDK.exists()) {
  AndroidUtil.createFileFromTemplate(emulatorPackageXMLTemplate, packageFileforEmulatorinSDK);
  System.out.println("Package.xml created in emulator");
}
```
There is a package.xml file to clone from the template. The issue is that there is no existing instance of mode here. 
[AndroidBuild.java](https://github.com/processing/processing-android/blob/475224e1d0572944f34306b9bb6dd4d1df8a92d0/mode/src/processing/mode/android/AndroidBuild.java#L156) is using mode's instance to use - `mode.getContentFile("templates/"...`
I must use the existing AndroidMode's instance somehow here to use `mode.getContentFile` in SDKDonwloader? Any suggestions on this would be helpful. Thanks!

**Comment by rupeshkumar22** [comment]
Found a way, Mode's instance [here](https://github.com/processing/processing-android/blob/55bfe5891be4b57b091553882eaf7a6062cd8876/mode/src/processing/mode/android/AndroidSDK.java#L489) can be passed to SDKDownloader's constructor [here](https://github.com/processing/processing-android/blob/55bfe5891be4b57b091553882eaf7a6062cd8876/mode/src/processing/mode/android/AndroidSDK.java#L491). By that way `mode.getContentFile("templates/"` can be used in SDKDownloader.

**Comment by codeanticode** [comment]
@rupeshkumar22 I actually not that convinced about manually creating the missing package.xml file for the emulator, seems like a problematic hack to me. I'd rather use the sdk manager to download the SDK as we discussed originally. I did just that in [this branch](https://github.com/processing/processing-android/tree/download_emu_on_demand). The idea is that the mode downloads the emulator only when the user tries to run a sketch in the emulator for the first time. I reused some of your code from this PR.

Seems to work well, at least on an intel Mac. On a M1 macbook, the emulator does not launch, but this is probably not related to the installation but to the new architecture.

**Comment by rupeshkumar22** [comment]
@codeanticode I have gone through your PR and the idea of downloading the emulator in `runInEmulatorHandler` seems the best. I understand the importance of using `sdkmanager` to download the emulator. I'm working on this PR that you have shared and testing it. Also, I am looking for ways to make it work on M1 architecture. It seems to be related to enabling of Hyper-V on M1.  Thanks for the direction. Let me know If you have any suggestions @ranaaditya

**Comment by rupeshkumar22** [comment]
Results from testing [download_emu_on_demand](https://github.com/processing/processing-android/tree/download_emu_on_demand) branch-
SDK and Emulator get installed properly. However, I see the following error during the first time launch
```
[==============================         ] 78% Unzipping... emulator/resources/ma
[=======================================] 100% Unzipping... emulator/resources/m

Output from emulator download 0
HAXM: Intel HAXM reinstalled successfully!
Emulator process exited with status 1.
EmulatorController: Emulator never booted. NOT_RUNNING
Error while starting the emulator. (NOT_RUNNING)
Shutting down any existing adb server... Done.
```
This error comes initially during the first time install of the emulator. Also, It was encountered previously. After that no issue was faced with `Run In Emulator options`. I'm looking into this issue. 
Thanks!

**Comment by rupeshkumar22** [comment]
@codeanticode @ranaaditya Above commits are cloned from the [download_emu_on_demand](https://github.com/processing/processing-android/tree/download_emu_on_demand) branch.

**Update on the above emulator issue on launching Emulator for the first time**-
It was strange that streampump was not providing the output streams from the process. However, On using RedirectStreamHandler I was able to get Error logs from the emulator launch process-
```
[==============================         ] 78% Unzipping... emulator/resources/ma
[=======================================] 100% Unzipping... emulator/resources/m

Output from emulator download 0
HAXM: Intel HAXM reinstalled successfully!
ERROR   | Unknown AVD name [processing-phone], use -list-avds to see valid list.
ERROR   | HOME is defined but there is no file processing-phone.ini in $HOME\.android\avd
ERROR   | (Note: Directories are searched in the order $ANDROID_AVD_HOME, $ANDROID_SDK_HOME\avd and $HOME\.android\avd)
INFO    | Android emulator version 31.2.10.0 (build_id 8420304) (CL:N/A)
Emulator process exited with status 1.
EmulatorController: Emulator never booted. NOT_RUNNING
Shutting down any existing adb server...Error while starting the emulator. (NOT_RUNNING)
 Done.
```
I'm exploring this issue further.
Thanks!

**Comment by codeanticode** [comment]
Something seems wrong with the AVD, after the first time you run the emulator, an AVD called processing-home should be created inside <path to sketchbook folder>/android/avd

**Comment by rupeshkumar22** [comment]
For me, the AVD getting created in `\Documents\Processing\android\avd\processing-phone`
It seems to be in sketchbook folder. 
Also I see some existing AVD directories present at home folder which might be created by Android Studio earlier. `C:\Users\Name\.android\avd\Galaxy_Nexus_API_28.avd`

**Comment by rupeshkumar22** [comment]
Also, I observe that after the first time emulator launch error only three files are present inside `\Documents\Processing\android\avd\processing-phone`
These files are-
`processing-phone/config.ini`
`processing-phone/sdcard.img`
`processing-phone/userdata.img`

**Comment by codeanticode** [comment]
That's correct, the AVD only comprises those three files. The actual system images are under sdk/system-images. I'd try deleting  ```\Documents\Processing\android\avd\processing-phone``` and running the emulator again,

**Comment by rupeshkumar22** [comment]
**Current progress on the emulator launch issue**-
Based on the error logs, I introduced the `$ANDROID_AVD_HOME` variable with the path as `<path to sketchbook folder>/android/avd` in the launch runtime process. After applying the changes the error logs show that it looks for `processing-phone.ini` inside the defined `$ANDROID_AVD_HOME` but it was not found. Only the `processing-phone` folder is there but beside it, `processing-phone.ini` file needs to be there.

I observe that during `avd create` command `processing-phone.ini` is getting generated in my home directory as follows-
`C:\Users\Name\.android\avd\processing-phone.ini`

Seems that it is the inbuilt behavior of generating `processing-phone.ini` at the Home directory as adding `$ANDROID_AVD_HOME` as environment in avd create process did not helped.

Moving processing-phone.ini to sketch's AVD folder and then using `Run In Emulator` works just fine. Looking for a way to generate `processing-phone.ini` at `<path to sketchbook folder>/android/avd` which resolves the first-time launch issue.
Any suggestions would be helpful.
Thanks!

**Comment by codeanticode** [comment]
Not sure what the problem is with your setup, there should no need to set an environmental variable for the mode to create the AVD, see the implementation [here](https://github.com/processing/processing-android/blob/master/mode/src/processing/mode/android/AVD.java#L359), where it uses the sketchbook folder to place android/avd.

**Comment by rupeshkumar22** [comment]
@codeanticode I have checked that `avdmanager create` part. I think the issue arises when `avdmanager create` command takes too long to create `processing-phone.ini` file and this file is being used during the emulator launch process. In that case, the emulator launch process exits with non-zero result code but that is a rare case that happened for me. We can align both processes in synch to resolve. If someone faces similar issues with non-zero return code by the emulator, I will work on this.

**Closing this PR** to create a new PR, as a lot of changes have been done in the official master branch and some part of RedirectStreamHandler of this PR has already been merged.

**Pull request closed without merging**